### ISSUE TITLE
Suggestions re modularity and multiple entry points for same image

### ISSUE THREAD

**Issue by payerle** [containers-ftw/apps#25]
I have been doing a bit with Singularity for the past year, but like the Sci-F %app* stuff in 2.4

I am currently looking to upgrade a keras/theano/tensorflow container image for my users using these features, but there are some aspects which do not seem to quite fit with the paradigm that is discussed in docs that I have read, which seem to stress workflow more than modularity.  I am willing to submit the resultant image to shub, but to do so it would be good to adhere to community standards.

Basically, I install some useful libs, python 2 and 3, and then python packages for each of keras, theano, and tensorflow for each python version.  (And eventually want to add cuda/nocuda options as well, but for now assume no cuda).  All three of the main packages (keras/theano/tensorflow) are
primarily just python modules (although one or two have an auxilliary command), and keras can
be called with an environmental variable to specify whether it should use theano or tensorflow as
a backend.

My "modular install" approach would have %appinstall, etc. entries for
python2
python3
theano-py2
tensorflow-py2
keras-py2
theano-py3
tensorflow-py3
keras-py3

However, this seems overkill from the "workflow" paradigm, since from workflow perspective basically
there are only python2 and python3, plus entry points for the theano/tensorflow backends of keras.

If I do something to do both modularity and workflow, there become so many entry points that it would seem to lose some of the Sci-F design goals.

Any advice?

**Comment by vsoch** [comment]
hey @payerle! I'd love to help with this! Could you post here (or send to me privately) what you have so far for install steps and I'll put something together for discussion? Just a heads up I'm going to be off the grid starting tomorrow for a bit, so I can try to get us going later today and we may need to pick up after that.

**Comment by vsoch** [comment]
If you want to come up with something, here is a quick question to help guide your design. I like to ask what is the primary purpose of your container? For example, when we think of workflows we often have a very particular application / action for the container to perform (such as running keras). So if you want this sort of use case, then you want to expose that (and we don't care so much about python, but we might make entrypoints to use keras in either version).  On the other hand, if your container is agnostic to workflows and wants to serve interactive environments (more of a development container) then you would likely want to expose the python 2/3 entrypoints (with their appropriate modules installed).

If you are thinking of extension of older containers (e.g., updated recipes to add support for new versions that maintain the old ones) then you have two things to think about. **Case 1**: you want the same entrypoint / command to be consistent across container runs (for example, let's say when some python 4 exists, running python 3 for a container that is updated with 4 will still run the same python 3). In this case you would have each of `keras-python3` and `keras-python4` so the commands are consistent indefinitely out into the future. **Case 2** you want the user to be able to issue the same command, but update the version over time. So now `keras-python` might point to `keras-python3` and later it would point to `keras-python4`.

How about then having both a `keras-python3` and `keras-python4` (someday, lol) and then a `keras-python` that simply activates the environment for whatever is the most recent version for the container, and executes the runscript? That way, executing `keras-python` will give the "latest" but the user can also decide to be specific about version and use `keras-python3/4`. Then for updating a container you would install new versions, and point the "latest" entrypoint to the latest :)


>> there are some aspects which do not seem to quite fit with the paradigm that is discussed in docs that I have read, which seem to stress workflow more than modularity.

If you want to ask about specific things, or something that is not clear, I can give more thinking about this. Overall, if you are primarily concerned with entrypoints for the different machine learning modules, and since python has good (isolated) environments (eg, miniconda2 and miniconda3) that you can easily add to the path for an app, I would install python2 and 3 in `%post` (and not on default path but in opt or similar) paths, and then add them to the $PATH for each of the apps (theanos, keras, etc). You could also have them be their own apps, but if the primary purpose of the container isn't for python it's probably fine to not. And then you would also export variables for each app to determine which backend to use. So the recipe would have a post something like:

```
%post
# updates, etc.
# install python2 and python3 as miniconda/anaconda somewhere off $PATH
```

and then for each of `theano-python2`, `theano-python3`, `tensorflow-python2`,`tensorflow-python3`, `keras-python2` and `keras-python3` you would have an environment, and an apprun (and maybe some labels :O)):

```
%appenv theano-python2
    PATH=/opt/anaconda2/bin:$PATH
    export PATH
    # backend variables

%apprun theano-python2
    exec /opt/anaconda2/bin/python "$@"
```
and then for your "latest" something like:

```
%apprun theano-python
    source $APPENV_theano_python3 
    exec $APPRUN_theano_python3
```

and actually, it wouldn't hurt to have python2 and python3 as apps given that we aren't installing to system anyhoo:

```
%appinstall python2
apt-get update && apt-get install -y python python-dev

%apprun python2
    exec /opt/anaconda3/bin/python "$@"

%appinstall python3
# same

%apprun python3
    exec /opt/anaconda3/bin/python "$@"
```

Now that I think of it, if we can accomplish the same thing using the modular approach, we should probably do that. Let me know if that helps!

**Comment by payerle** [comment]
Thanks for your input.  I think I was trying to use SciF apps to address
the wrong problem.

I am a sysadmin, not an "application user", and while I certainly want any
container I build to be easily usable by my users, I was also looking ease
of maintaining these containers (unfortunately, the "ideal" of having my
users maintain their own container has not materialized yet:).

I was looking at using the %app* sections to isolate the installation
process for various components in the image, even if such components were
not pieces that would be exposed to the workflow.  I see now that is an
abuse of the SciF %app* stuff.  (And it appears unneeded, as some quick
tests seem to indicate that one can have multiple %setup, %post, etc.
sections, at least in 2.4.)

So I will restrict the use of SciF %app* stuff for the entry points
that should be exposed to the user.

Thanks for your feedback.

On Wed, 10 Jan 2018, Vanessa Sochat wrote:

> 
> If you want to come up with something, here is a quick question to help guide your design. I like to ask what is the primary purpose of your container? For example, when we think of workflows we often have a very
> particular application / action for the container to perform (such as running keras). So if you want this sort of use case, then you want to expose that (and we don't care so much about python, but we might make
> entrypoints to use keras in either version). On the other hand, if your container is agnostic to workflows and wants to serve interactive environments (more of a development container) then you would likely want to expose
> the python 2/3 entrypoints (with their appropriate modules installed).
> 
> If you are thinking of extension of older containers (e.g., updated recipes to add support for new versions that maintain the old ones) then you have two things to think about. Case 1: you want the same entrypoint /
> command to be consistent across container runs (for example, let's say when some python 4 exists, running python 3 for a container that is updated with 4 will still run the same python 3). In this case you would have each
> of keras-python3 and keras-python4 so the commands are consistent indefinitely out into the future. Case 2 you want the user to be able to issue the same command, but update the version over time. So now keras-python might
> point to keras-python3 and later it would point to keras-python4.
> 
> How about then having both a keras-python3 and keras-python4 (someday, lol) and then a keras-python that simply activates the environment for whatever is the most recent version for the container, and executes the
> runscript? That way, executing keras-python will give the "latest" but the user can also decide to be specific about version and use keras-python3/4. Then for updating a container you would install new versions, and point
> the "latest" entrypoint to the latest :)
>
>             there are some aspects which do not seem to quite fit with the paradigm that is discussed in docs that I have read, which seem to stress workflow more than modularity.
> 
> If you want to ask about specific things, or something that is not clear, I can give more thinking about this. Overall, if you are primarily concerned with entrypoints for the different machine learning modules, and since
> python has good (isolated) environments (eg, miniconda2 and miniconda3) that you can easily add to the path for an app, I would install python2 and 3 in %post (and not on default path but in opt or similar) paths, and then
> add them to the $PATH for each of the apps (theanos, keras, etc). You could also have them be their own apps, but if the primary purpose of the container isn't for python it's probably fine to not. And then you would also
> export variables for each app to determine which backend to use. So the recipe would have a post something like:
>
>  %post
> # updates, etc.
> # install python2 and python3 as miniconda/anaconda somewhere off $PATH
> 
> and then for each of theano-python2, theano-python3, tensorflow-python2,tensorflow-python3, keras-python2 and keras-python3 you would have an environment, and an apprun (and maybe some labels :O)):
>
>  %appenv theano-python2
>     PATH=/opt/anaconda2/bin:$PATH
>     export PATH
>     # backend variables
> 
> %apprun theano-python2
>     exec /opt/anaconda2/bin/python "$@"
> 
> and then for your "latest" something like:
>
>  %apprun theano-python
>     source $APPENV_theano_python3
>     exec $APPRUN_theano_python3
> 
> and actually, it wouldn't hurt to have python2 and python3 as apps given that we aren't installing to system anyhoo:
>
>  %appinstall python2
> apt-get update && apt-get install -y python python-dev
> 
> %apprun python2
>     exec /opt/anaconda3/bin/python "$@"
> 
> %appinstall python3
> # same
> 
> %apprun python3
>     exec /opt/anaconda3/bin/python "$@"
> 
> Now that I think of it, if we can accomplish the same thing using the modular approach, we should probably do that. Let me know if that helps!
> 
> ?
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.[ASZO_zNX24UznURXqPnqRzF2LDTg3uULks5tJQcNgaJpZM4RZl1m.gif]
> 
> 
>

Tom Payerle
DIT-ACIGS/Mid-Atlantic Crossroads 	payerle@umd.edu
5825 University Research Court 		(301) 405-6135
University of Maryland
College Park, MD 20740-3831

**Comment by vsoch** [comment]
No problem! Note that I think this is a very good use case for scif, in that an entire "app" doesn't necessarily have to be an application, it's an environment / install that can be chosen for use (and to the user, perhaps, this subtle change does map to a different application") Please re-open if you want to discuss further!
